src/realm/sync/client.cpp

```diff
@@ -740,49 +743,71 @@ void ClientImpl::remove_connection(ClientImpl::Connection& conn) noexcept
 
 void SessionImpl::force_close()
 {
-    m_wrapper.force_close();
+    // Allow force_close() if session is active or hasn't been activated yet.
+    if (m_state == SessionImpl::Active || m_state == SessionImpl::Unactivated) {
+        m_wrapper.force_close();
+    }
 }
 
 void SessionImpl::on_connection_state_changed(ConnectionState state,
                                               const util::Optional<SessionErrorInfo>& error_info)
 {
-    m_wrapper.on_connection_state_changed(state, error_info); // Throws
+    if (m_state == SessionImpl::Active) {
+        m_wrapper.on_connection_state_changed(state, error_info); // Throws
+    }
 }
 
 
 const std::string& SessionImpl::get_virt_path() const noexcept
 {
+    // Can only be called if the session is active or being activated
+    REALM_ASSERT(m_state == State::Active || m_state == State::Unactivated);
     return m_wrapper.m_virt_path;
 }
 
 const std::string& SessionImpl::get_realm_path() const noexcept
 {
+    // Can only be called if the session is active or being activated
+    REALM_ASSERT(m_state == State::Active || m_state == State::Unactivated);
```

Comment:
Really great to see these extra assertions! Not blocking, but where you can, I prefer to add a bit of info to help debugging in case they are violated.
```suggestion
    REALM_ASSERT_EX(m_state == State::Active || m_state == State::Unactivated, m_state);
```

Reply:
Thank you! I agree it will be nice to have better debugging. I will update these.